### DISCUSSION TITLE
[@types/react] would type PropsWithoutRef<P> work better with an Omit

### DISCUSSION THREAD

**Discussion by johnslemmer** [DefinitelyTyped/DefinitelyTyped#64163] [Issues with a @types package] [answered]
I'm having this issue: https://github.com/expo/router/issues/236.  But to sum up, when types are emitted using the `forwardRef`/`PropsWithoutRef` type it is very restrictive because of the long `Pick` list that gets generated.  I'm wondering if `Omit` would be a better tool for the job.  Maybe changing [the PropsWithoutRef type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/986075d3be84d3f35dc2b00c6dad8ede2e5f1974/types/react/index.d.ts#L784) to look more like this:

```ts
    type PropsWithoutRef<P> =
        P extends any ? ('ref' extends keyof P ? Omit<P, 'ref'> : P) : P;
```

Would this still work and solve the issue of this massive and restrictive `Pick` list being generated?

[a comment by an automated account is not shown]

**Comment by johnslemmer** [comment]
I think my suggestion is redundant.  Any reason it can't simply be:

```ts
    type PropsWithoutRef<P> = Omit<P, 'ref'>
```

**Reply by dimitropoulos** [reply]
It can't be just Omit because it this type, `PropsWithoutRef`, allows a _union_ of `P`s.  If it were changed like you suggest it would still work fine for one type, but would break down for unions.

I don't think it would make much difference to add `Omit` in the place you mentioned.  The part that's critical is `P extends any` which forces the distributive conditional.

here's a playground to explain:

https://www.typescriptlang.org/play?ssl=2&ssc=13&pln=2&pc=16#code/C4TwDgpgBACgTgezAZwOoEtgAsEFdgBKEAZgDwwB8UAvAFBSxQQAewEAdgCbJQCG7IelAD8UAORwSYpqw7coAawggExWEIaiY6AMYLyAGigBRZjoA2uThFJKVamEYlSKFDVABc6hl5i1aoJCwiCgY2HiEJADyALaY5FTUULHxjuKSxGJuAeDQAMIIMWAI7BzAAII0UADefF4AjEYZXmK8RJlQAL45QQBSuMjAUaVV8EhomDj47aQFRSVl5W4A9MsMUAB6wj3Q-YPDECnAoyET4dPRccCzhcWl7BUrawxb-oH5twsPAEJVtQBGXgATE0SC1-u1pJ0ANxvXJQOZ3Mo8JKIr4VKAAHwRn3uwG+sJ2UAAqux0CUTuMwlNImQ0XjkE91q93iSySUjpTQpMIu0jjd5gymS9hEA

**Reply by johnslemmer** [reply]
Awesome, thanks for the explanation!

Ok, and I have tested this by manually changing react types in the expo-router.  And the subsequent build emits a much cleaner type.

And I have also tested that an Omit used in my own project allows for the behavior I'm expecting.

I'm going to go ahead and open a PR.

**Comment by johnslemmer** [comment, marked as answer]
PR open: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64166

**Reply by johnslemmer** [reply]
PR was merged.  Thanks for the help!
